### ISSUE TITLE
Premultiplication, if any, must occur before shrink

### ISSUE THREAD

**Issue by CmdrShepardsPie** [lovell/sharp#605]
We're using Sharp for a project that will resize a PNG/24 image with a transparent background to many different sizes; and we're finding that it's calculating the "blending" between the edge of the opaque image and the transparent background to include the "color" of the "transparency" (whether it's RGBA(255, 255, 255, 0) or RGBA(0, 0, 0, 0)) causing an edge or halo to occur. I can confirm this by making the background transparency white (white halo) or black (black halo). I believe this is a bug in the raster resizing because resizing it with a program doesn't introduce this. Attached are the examples (original image had a white "O" with a "black" transparent background, other is the inverse; both were then put over a white and black background respectively to show the outlining issue.)

![o-black](https://cloud.githubusercontent.com/assets/22330942/19491218/5eecaa3a-952f-11e6-8ba7-cc56936341a1.png)
![o-white](https://cloud.githubusercontent.com/assets/22330942/19491217/5ee12746-952f-11e6-849f-47758834ab88.png)

**Comment by lovell** [comment]
Hello, are you able to share the original images (preferably not via Github as it will process them), and ideally also the code that will generate this output?

**Comment by CmdrShepardsPie** [comment]
Images:
https://www.dropbox.com/sh/n2e6sri7fhqlffv/AAD78bojAg5CjvjSTl4wOKWja?dl=0

The code is really simple, basically just:
request.get(url)...on(accumulate data)...then(return as buffer)
var pipeline = sharp(image.buffer);
pipeline = pipeline.clone(); // we're resizing into multiple sizes
pipeline = pipeline.resize(imageOutput.width, imageOutput.height);
pipeline.toBuffer(return new buffer)
aws upload file, store buffer as body

**Comment by lovell** [comment]
Thanks Chris, how were the original images created?

The metadata in the header `Png:IHDR.bit-depth-orig: 8` states this image uses 8 bits (0-255) but the pixel values themselves appear to be 16 bits (0-65535).

```
$ identify -verbose o-white-in.png
Image: o-white-in.png
  Format: PNG (Portable Network Graphics)
  Geometry: 510x500
  Class: DirectClass
  Type: bilevel
  Depth: 1 bits-per-pixel component
  Channel Depths:
    Gray:     1 bits
    Opacity:  8 bits
  Channel Statistics:
    Gray:
      Minimum:                     0.00 (0.0000)
      Maximum:                 65535.00 (1.0000)
      Mean:                    30632.34 (0.4674)
      Standard Deviation:      32697.93 (0.4989)
    Opacity:
      Minimum:                     0.00 (0.0000)
      Maximum:                 65535.00 (1.0000)
      Mean:                    35160.22 (0.5365)
      Standard Deviation:      32594.40 (0.4974)
  Opacity: (  0,  0,  0,255)      #000000FF
  Filesize: 9.2Ki
  Interlace: No
  Orientation: Unknown
  Background Color: white
  Border Color: #DFDFDF00
  Matte Color: #BDBDBD00
  Page geometry: 510x500+0+0
  Compose: Over
  Dispose: Undefined
  Iterations: 0
  Compression: Zip
  Png:IHDR.color-type-orig: 6
  Png:IHDR.bit-depth-orig: 8
  Signature: b34d87c3ec2778fa6142e6f878cc9e0f72ae6cdb9e1723a3d562d2bf919dbe3b
  Tainted: False
```

(In addition, I believe `Png:IHDR.color-type-orig: 6` is RGBA rather than greyscale+alpha so something appears to be quite wrong here.)

The logic in libvips to work out the pixel format interpretation is selecting `VIPS_INTERPRETATION_sRGB` for this image.

https://github.com/jcupitt/libvips/blob/master/libvips/foreign/vipspng.c#L310-L321

This means sharp will consider pixel values to be in the 0-255 range and therefore incorrectly clip values when (un)premultiplying, hence the ringing effect.

**Comment by CmdrShepardsPie** [comment]
They were provided to my team as AI files, and one of us exported it as a PNG from Illustrator, and these images specifically were then cropped by Photoshop and exported as PNGs (all with same results).

**Comment by lovell** [comment]
Photoshop has historically been a big offender when it comes to broken output (wrong bit depth, mismatched colour profiles etc.).

Perhaps try setting/forcing the image format to 8 bits per channel in Photoshop before exporting then use ImageMagick's `identify -verbose filename.png` tool to verify the channel statistics are within the 0-255 range.

Good luck!

**Comment by lovell** [comment]
@CmdrShepardsPie Were you able to make any progress exporting valid PNG images from Photoshop?

**Comment by teroparvinen** [comment]
I am having a similar problem. A very simple resize operation to size images to half the size:

```
var image = sharp(fileName);
image
	.metadata()
	.then((metadata) => {
		return image
			.resize(Math.round(metadata.width / 2))
			.toBuffer((err, buffer, info) => {

				// Save buffer

			});
	});
}));
````

Example files, source and result are attached.
![pensas](https://cloud.githubusercontent.com/assets/1235386/20709622/213d2a4e-b63f-11e6-8b2e-1b6a13412f6e.png)
![pensas 2x](https://cloud.githubusercontent.com/assets/1235386/20709623/213d40b0-b63f-11e6-8634-b81623dde3c1.png)

**Comment by lovell** [comment]
@teroparvinen In your example, a shrink of exactly 1/2, a fast box shrink is used which currently does not premultiply pixel values first. Moving the premultiply operation to occur sooner in the pipeline and therefore be used for box shrinking should help but I'll need to experiment with this. Thanks for reporting here.

Both these problems appear to relate to incorrect premultiplication but I think different code paths are behind them.

**Comment by lovell** [comment]
Commit d3c78f8 on the `master` branch ensures the pixel premultiplication step occurs before any shrink operation, which removes the dark edges when I tested locally with @teroparvinen's example.

A few of the test expectations needed updating as the quality improvements broke the thresholds (a good thing). It also exposed a previously undetected bug in one of the "smart" crop strategies, so a rather useful fix all round, thank you.

@CmdrShepardsPie Are you able to retest with your original images?

**Comment by lovell** [comment]
This improvement is now available in v0.17.0 (but please do feel free to re-open if you're still experiencing problems).
